compiler/src/iree/compiler/DispatchCreation/InsertTensorBarriers.cpp

```diff
@@ -0,0 +1,144 @@
+#include "iree/compiler/DispatchCreation/Passes.h.inc"
+
+namespace {
+
+// Check if an operation is a compute operation (tensor/linalg/linalgext).
+static bool isComputeOp(Operation *op) {
+  if (!op) {
+    return false;
+  }
+  auto *dialect = op->getDialect();
+  if (!dialect) {
+    return false;
+  }
+  return isa<linalg::LinalgDialect, IREE::LinalgExt::IREELinalgExtDialect,
+             tensor::TensorDialect>(dialect);
+}
+
+// Traverse forward along use-def chains starting from `val` to identify values
+// that flow into compute operations. These values are candidates for inserting
+// barrier.start operations.
+static void collectInputsToComputeRegion(Value val,
+                                         llvm::SetVector<Value> &inputValues,
+                                         llvm::DenseSet<Value> &visited) {
+  if (!visited.insert(val).second) {
+    return;
+  }
+
+  for (OpOperand &use : val.getUses()) {
+    Operation *userOp = use.getOwner();
+    if (!userOp) {
+      continue;
+    }
+    if (isComputeOp(userOp)) {
+      inputValues.insert(val);
+    } else {
+      for (Value result : userOp->getResults()) {
+        collectInputsToComputeRegion(result, inputValues, visited);
+      }
+    }
+  }
+}
+
+// Traverse backward along use-def chains starting from `val` to identify values
+// produced by compute operations. These values are candidates for inserting
+// barrier.end operations.
+static void
+collectOutputsFromComputeRegion(Value val, llvm::SetVector<Value> &outputValues,
+                                llvm::DenseSet<Value> &visited) {
+  Operation *definingOp = val.getDefiningOp();
+  if (!definingOp || !visited.insert(val).second) {
+    return;
+  }
+
+  if (isComputeOp(definingOp)) {
+    outputValues.insert(val);
+  } else {
+    llvm::for_each(definingOp->getOperands(), [&](Value operand) {
+      collectOutputsFromComputeRegion(operand, outputValues, visited);
+    });
+  }
+}
+
+struct InsertTensorBarriersPass final
+    : public impl::InsertTensorBarriersPassBase<InsertTensorBarriersPass> {
+  using Base::Base;
+
+  void runOnOperation() override {
+    auto funcOp = getOperation();
+    OpBuilder builder(funcOp.getContext());
+
+    // Insert barrier.start operations for values that flow into compute ops.
+    llvm::SetVector<Value> needsStartBarrier;
+    llvm::DenseSet<Value> visited;
+    llvm::for_each(funcOp.getArguments(), [&](BlockArgument arg) {
+      collectInputsToComputeRegion(arg, needsStartBarrier, visited);
+    });
+
+    for (Value val : needsStartBarrier) {
+      if (!isa<RankedTensorType>(val.getType())) {
+        continue;
+      }
+      builder.setInsertionPointAfterValue(val);
+      auto startOp =
+          IREE::TensorExt::BarrierStartOp::create(builder, val.getLoc(), val);
+
+      val.replaceUsesWithIf(startOp.getResult(), [&](OpOperand &use) {
+        return isComputeOp(use.getOwner()) &&
+               !isa<tensor::DimOp>(use.getOwner());
+      });
+    }
```

Comment:
nit: I'd remove the blank line because they belong to the same code block to me. I.e., I don't see benefits from having a blank line.

https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace